### DISCUSSION TITLE
How to change primary background color of the main window for FlatLaf Light?

### DISCUSSION THREAD

**Discussion by PavelTurk** [apache/netbeans#4282] [Q&A] [answered]
I want to make default light FlatLat theme a little darker. I tried to modify `netbeans-14/netbeans/platform/modules/org-netbeans-swing-laf-flatlaf.jar#org/netbeans/swing/FlatLaf.properties file`, but it seems that all modifications in this file are ignored.  For example,
I set:

```
nb.multitabs.background=#ff0000
nb.multitabs.foreground=#ff0000
nb.multitabs.activeBackground=#ff0000
nb.multitabs.selectedBackground=#ff0000
nb.multitabs.selectedForeground=#ff0000
nb.multitabs.hoverBackground=#ff0000
nb.multitabs.underlineColor=#ff0000
nb.multitabs.inactiveUnderlineColor=#ff0000
```

And there are no changes in the appearance. Could anyone say how to do it?

**Comment by DevCharly** [comment, marked as answer]
Set `@background` in `platform/o.n.swing.laf.flatlaf/src/org/netbeans/swing/laf/flatlaf/FlatLightLaf.properties`.

E.g.
~~~
@background = #e0e0e0
~~~

would change the look to:

![image](https://user-images.githubusercontent.com/5604048/176389840-34e8095e-69fa-4bf8-8fc4-db15c9a102cb.png)

**Reply by PavelTurk** [reply]
Do I understand you correctly - I need to set this property in source code and after build NB? Because I tried to set it in binary `org-netbeans-swing-laf-flatlaf.jar` file and color didn't change.

**Reply by DevCharly** [reply]
> Do I understand you correctly - I need to set this property in source code and after build NB?

No, `FlatLightLaf.properties` is a simple text file. There is no need to compile it.

Changing it in jar should work

**Reply by PavelTurk** [reply]
But I can't make it work. This is my prop file:

![Screenshot from 2022-07-02 13-18-08](https://user-images.githubusercontent.com/10261361/176996413-17126ab0-6626-44df-87d3-a0922c2977ff.png)

And this is the result:

![Screenshot from 2022-07-02 13-18-14](https://user-images.githubusercontent.com/10261361/176996421-0f93501f-e0c3-406e-95c8-a3b0dfd77348.png)

Could you say what can be wrong here?

**Reply by DevCharly** [reply]
That is surprising 😕 

Changing in Jar indeed does not work, but it works when doing same change in IDE and start NB in debugger...

![image](https://user-images.githubusercontent.com/5604048/177002571-87677aed-6e0e-4c9e-bb27-2f925ad6d154.png)

Don't know why? There should be no difference...

**Reply by neilcsmith-net** [reply]
@DevCharly does clearing the cache folder make that work? I was wondering whether we update to support additional properties in user config too.

**Reply by DevCharly** [reply]
> does clearing the cache folder make that work? 

Yes, removing cache helps. Many thanks.
I've removed `C:\Users\myusername\AppData\Local\NetBeans\Cache\14`.
Does NB copy properties files to the cache?

> I was wondering whether we update to support additional properties in user config too.

That would indeed make such customization easier.
Adding following (e.g. to `org.netbeans.swing.laf.flatlaf.Installer`) should work (not tested):

~~~java
File configDir = new File( ... );
if( configDir.isDirectory() ) {
    FlatLaf.registerCustomDefaultsSource( configDir.toURI().toURL() );
}
~~~

**Reply by DevCharly** [reply]
Update: there is also a register method that takes a `File`:

~~~java
File configDir = new File( ... );
FlatLaf.registerCustomDefaultsSource( configDir );
~~~

FlatLaf then searches for `FlatLaf.properties`, `FlatLightLaf.properties` and `FlatDarkLaf.properties` in those directory.

**Reply by PavelTurk** [reply]
@DevCharly Yes, thank you. Now `@background` color changes. Only one more question - how to set background color for project/files/navigator/services main field? I think it is somehow linked to `@background` color because it became darker too. I tried to change all properties using `@background` but didn't find.

**Reply by DevCharly** [reply]
> how to set background color for project/files/navigator/services main field?

I think `Tree.background` should work.

**Reply by PavelTurk** [reply]
@DevCharly  and @neilcsmith-net , thank you very much for your help!

**Reply by neilcsmith-net** [reply]
> Does NB copy properties files to the cache?

@DevCharly IIRC classes and resources, at least from certain JARs, are cached - see https://github.com/apache/netbeans/blob/master/platform/o.n.bootstrap/arch.xml#L1074
